GeochemistryUnitConverter utility added, along with unit tests

WilkAndy opened the pull request:
Refs #17155

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/710374/) on 6570385 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17343/site)

This comment will be updated on new commits.

WilkAndy committed:
GeochemistryUnitConverter utility added, along with unit tests
Refs #17155

WilkAndy committed:
Add notion of units and a new constraint_meaning interface to GeochemicalSystem

The old constraint system had unit choices built in, so with this new unit functionality, the constraint_meaning names did not make sense.  Hence, a new user interface for the constraint_meaning has been created.  It is both simpler and more flexible because units and meanings are now different things, rather than all lumped together.  The new user-interface constraint_meanings are mapped into the old constraint_meanings, during the constructor of GeochemicalSystem, at the same time that the units are all converted to mole-based units

Refs #17155

WilkAndy committed:
tests and documentation of the new geochemistry units stuff

At the same time, various old tests and documentation has been modified to use the new units system.

At the same time, the new HIT notation ([] instead of [../], etc) has been used in all geochemistry tests

Fixes #17155

WilkAndy commented:
Hey @loganharbour , i'm looking for a review on this.   Don't despair - the vast majority of the 4000 lines is new tests and re-doing the doco and old tests.

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/711783/) on 26307b0 : invalidated by @WilkAndy

unusual error: /opt/civet/build_0/moose/modules/ray_tracing/lib/libray_tracing-dbg.so: file not recognized: File truncated

WilkAndy commented:
Just reminding @loganharbour about this PR

loganharbour commented on modules/geochemistry/include/utils/GeochemistryUnitConverter.h:
```diff
@@ -0,0 +1,101 @@
+ */
+namespace GeochemistryUnitConverter
+{
+enum class GeochemistryUnits
```
```suggestion
enum class GeochemistryUnit
```

We usually keep these singular - do so if you desire consistency :)

WilkAndy replied:
Sure, done.

loganharbour commented on modules/geochemistry/include/utils/GeochemistryUnitConverter.h:
```diff
@@ -0,0 +1,101 @@
+ * species_name
+ */
+Real fromMoles(Real moles,
+               const GeochemistryUnits & unit,
```
```suggestion
               const GeochemistryUnits unit,
```

FYI with something like this enum that is an int, having this as a reference could be slower! There's a few others below.

WilkAndy replied:
Aha, i get it, thanks.

loganharbour reviewed (commented):
Looks good to me. Two small things that you could change if you desire.

We really do need a general unit framework somewhere in the framework.... but it doesn't exist yet!

WilkAndy committed:
@loganharbour changes.  Fixes #17155

WilkAndy commented:
Hey @loganharbour - i've implemented your changes.  Let's hope we see all green soon.

WilkAndy commented:
Thanks @loganharbour for the review and merge
